### PR TITLE
add optional ruby-profiling with --profile-ruby

### PR DESCRIPTION
**Pull request by lamont-granquist**
dumps a large call graph into /var/chef/cache/graph_profile.out

### CODE REVIEW

**Comment by lamont-granquist** [comment]
I've been using this a lot lately.  Seems it would be good to get it into PR shape and let other people play with it.

@chef/client-core review?

File: lib/chef/client.rb [hunk 1]
```diff
@@ -53,6 +53,11 @@
 require 'chef/mixin/deprecation'
 require 'ohai'
 require 'rbconfig'
+begin
+  require 'ruby-prof'
```
Comment (jaym) [inline lib/chef/client.rb]: why not do this in start_profiling or profiling_prereqs

**Comment by coderanger** [comment]
I feel like maybe this shouldn't show in the `--help` output as it is likely to be a red herring for users that want to profile their recipe code? Or at least if we want this to be a part of user workflows, we would need some tools to make it easier to filter user code vs. core code, etc.

**Comment by jaym** [comment]
:+1: huge fan

**Comment by ranjib** [comment]
agree with @coderanger , given chef-client does not have dependency on ruby-prof, it feels strange to expose this as CLI flag. a config option should be good enough
:+1:

**Comment by jaym** [comment]
Are there any downsides to making it a dependency? I've had to walk people through profiling chef runs and having this just work would be awesome. I agree that it's not the most perfect feature, but it's still very useful.

**Comment by coderanger** [comment]
@jaym I would say it would fall under the kind of thing to include with the omnibus packages but not actually be a dependency (for those of us left that still install via gems).

**Comment by lamont-granquist** [comment]
Its a dev dependency for that reason.  'gem install chef' won't pull it in, but 'bundle install' (not --without deployment) will, so it'll get shipped in omnibus.

And whats the API to hide it from --help?

File: lib/chef/client.rb (hunk 1 above)
Comment (lamont-granquist) [inline lib/chef/client.rb]: 6 of one, half dozen of the other?  I certainly can...

**Comment by danielsdeleo** [comment]
There isn't a way to hide things from help in mixlib-cli (or else we would have done this with some other flags we added just for tests). I suppose you could detect whether the gem is available at file load time and then surround the `option` call with an if statement. Otherwise you could add more explanatory text to the description.

**Comment by lamont-granquist** [comment]
If I omit the description it still shows up in --help.  I really want it a CLI option because temporarily adding it to /etc/chef/client.rb is fussy, particularly if you have chef managing client.rb and have to edit it on every run.  I avoided adding it as a config option, since I don't think people will want to have this option running on every run, particularly since the outfile gets truncated and its not a log (and that seems like a good way to DoS your disk space if we made it a log since the call graph is huge).

**Comment by lamont-granquist** [comment]
yeah, i could `if defined?` it, but if the gem is shipped with the omnibus packages, then it'll load for nearly everyone anyway, which defeats the purpose...

**Comment by danielsdeleo** [comment]
I guess it depends on the purpose. If we just don't want to make a hard dep on the profiler gem, that'd be fine, if we want only expose this to users who know it's there, then it wouldn't help.

**Comment by thommay** [comment]
I definitely agree this makes much more sense as a CLI flag than a config option. If you make with some typey-typey to make it really obvious what this does in the description I think we're good.

**Comment by danielsdeleo** [comment]
I'm in agreement with @thommay's suggestion.

**Comment by coderanger** [comment]
Yeah, not worth a ton of work to add hidden CLI args, but something to warn off peoples would be nice. Also would be good to move the require I think so it doesn't add to the load time for the 99% of users that don't need it (but will have the gem thanks to omnibus et al). That's a minor issue though, a few more requires aren't going to be super noticeable.

**Comment by lamont-granquist** [comment]
Made the description substantially more scary.

I'll move the load in a sec, lazying the load time does marginally win in my mind over grouping the requires together.

**Comment by coderanger** [comment]
:+1:

File: lib/chef/client.rb [hunk 2]
```diff
@@ -891,6 +896,30 @@ def run_failed_notifications
     attr_reader :override_runlist
     attr_reader :specific_recipes
 
+    def profiling_prereqs!
+      begin
+        require 'ruby-prof'
+      rescue LoadError
+        raise "You must have the ruby-prof gem installed in order to use --profile-ruby"
+      end
```
Comment (thommay) [inline lib/chef/client.rb]: as a truly trivial nitpick, the `begin` is redundant here no?

**Comment by thommay** [comment]
I'm :+1: regardless of my comment

File: lib/chef/application/solo.rb [hunk 3]
```diff
@@ -52,6 +52,12 @@ class Chef::Application::Solo < Chef::Application
     :boolean      => true,
     :default      => false
 
+  option :profile_ruby,
+    :long         => "--[no-]profile-ruby",
+    :description  => "Output ruby execution profile graph",
```
Comment (danielsdeleo) [inline lib/chef/application/solo.rb]: Might want to copy the scary here, too?

**Comment by danielsdeleo** [comment]
Probably want to update the description for the CLI option for solo and apply also. :+1: aside from that.

**Pull request merged**